safe_control_gym/safety_filters/cbf/cbf.py

```diff
@@ -0,0 +1,327 @@
+'''Control barrier function (CBF) quadratic programming (QP) safety filter.
+
+Reference:
+    * [Control Barrier Functions: Theory and Applications](https://arxiv.org/abs/1903.11199)
+'''
+
+import numpy as np
+import casadi as cs
+
+from safe_control_gym.safety_filters.base_safety_filter import BaseSafetyFilter
+from safe_control_gym.safety_filters.cbf.cbf_utils import linear_function, cbf_cartpole, cartesian_product
+
+
+class CBF(BaseSafetyFilter):
+    '''Control Barrier Function Class. '''
+
+    def __init__(self,
+                 env_func,
+                 slope: float=0.1,
+                 soft_constrained: bool=True,
+                 slack_weight: float=10000.0,
+                 slack_tolerance: float=1.0E-3,
+                 **kwargs):
+        '''
+        CBF-QP Safety Filter: The CBF's superlevel set defines a positively control invariant set.
+        A QP based on the CBF's Lie derivative with respect to the dynamics allows to filter arbitrary control
+        inputs to keep the system inside the CBF's superlevel set.
+
+        Args:
+            env_func (partial BenchmarkEnv): Functionalized initialization of the environment.
+            slope (float): The slope of the linear function in the CBF.
+            soft_constrainted (bool): Whether to use soft or hard constraints.
+            slack_weight (float): The weight of the slack in the optimization.
+            slack_tolerance (float): How high the slack can be in the optimization.
+        '''
+
+        # TODO: Combine with CLF approach for stabilization
```

Comment:
Should we remove these todos? And open new issues for each of these to track?

Reply:
@JacopoPan general question on structure: would you rather leave the TODOs in (making it look somewhat unfinished), open an issue, or some other way of tracking these issues?

Reply:
Both TODOs and issues are good.